src/app/static/src/app/store/adrUpload/actions.ts

```diff
@@ -0,0 +1,117 @@
+                                `${project.name}_naomi_outputs.zip`,
+                                "uploadFileOutputZip"),
+                            outputSummary: constructUploadFile(
+                                metadata,
+                                1,
+                                schemas.outputSummary,
+                                `${project.name}_naomi_summary.html`,
+                                "uploadFileOutputSummary")
+                        };
+
+                        commit({type: ADRUploadMutation.SetUploadFiles, payload: uploadFiles});
+                    }
+                });
+        }
+    },
+
+    async uploadFilesToADR(context, uploadFilesPayload) {
+        const {rootState, commit, dispatch} = context;
+        const uploadMetadata = rootState.modelRun.result?.uploadMetadata
+        const selectedDatasetId = rootState.baseline.selectedDataset!.id;
+        const modelCalibrateId = rootState.modelCalibrate.calibrateId;
+
+        commit({type: ADRUploadMutation.ADRUploadStarted, payload: uploadFilesPayload.length});
+
+        for (let i = 0; i < uploadFilesPayload.length; i++) {
+            commit({type: ADRUploadMutation.ADRUploadProgress, payload: i + 1});
+            const {resourceType, resourceFilename, resourceId} = uploadFilesPayload[i]
+
+            const requestParams: Dict<string> = {resourceFileName: resourceFilename}
+            if (resourceId) {
+                requestParams["resourceId"] = resourceId
+            }
+            if (resourceType === rootState.adr.schemas?.outputSummary) {
+                requestParams["description"] = uploadMetadata
+                    ? uploadMetadata.outputSummary.description
+                    : "Naomi summary report uploaded from Naomi web app"
+            }
+            if (resourceType === rootState.adr.schemas?.outputZip) {
+                requestParams["description"] = uploadMetadata
+                    ? uploadMetadata.outputZip.description
+                    : "Naomi output uploaded from Naomi web app"
+            }
+
+            let apiRequest = api<ADRUploadMutation, ADRUploadMutation>(context)
+                .withError(ADRUploadMutation.SetADRUploadError);
+            if (i === uploadFilesPayload.length - 1) {
+                apiRequest = apiRequest.withSuccess(ADRUploadMutation.ADRUploadCompleted);
+            } else {
+                apiRequest = apiRequest.ignoreSuccess();
+            }
+            const response = await apiRequest.postAndReturn(`/adr/datasets/${selectedDatasetId}/resource/${resourceType}/${modelCalibrateId}`,
+                qs.stringify(requestParams))
+            if (!response) {
+                break
+            }
+        }
+        await getAndSetDatasets(context, selectedDatasetId)
+        dispatch("getUploadFiles");
+    }
+};
+
+async function getAndSetDatasets(context: ActionContext<ADRUploadState, RootState>, selectedDatasetId: string) {
```

Comment:
Absolutely agree - bizarrely I have no recollection of copy-pasting but it definitely shouldn't be duplicated!

What you suggested seems simple enough so I've done that but it feels a bit hacky. So before I continue to fix the tests, do you think it make more sense for `getAndSetDatasets()` to be a full-blown action within the `adr` module? It would only be invoked by other actions (from `adr` and `adrUpload`), and it might make the tests a bit more complicated, but it seems to fit the general pattern for an action of "using the context to make an API call and then doing a mutation".

Reply:
Yes, I think it definitely could be an action, manually dispatched from another action. I think it's fine to just `await` that dispatch if the action needs its child action to complete before it proceeds.